File: backend/api/models/ephemeris.py
"""
Pydantic models for ephemeris API endpoints.

These models define the request and response schemas for the ephemeris API,
ensuring type safety and validation for planetary position calculations.
"""

from datetime import datetime
from typing import List

from pydantic import BaseModel, Field


class PlanetPosition(BaseModel):
    """Position data for a planetary body."""

    position: float = Field(..., description="Position in degrees")
    retrograde: bool = Field(
        ..., description="Whether the planet is retrograde"
    )


class CalculationRequest(BaseModel):
    """Request model for planetary position calculation."""

    julian_day: float = Field(..., description="Julian Day Number")
    planet: str = Field(
        ..., description="Planet name (e.g., 'sun', 'moon', 'mercury')"
    )


class CalculationResponse(BaseModel):
    """Response model for planetary position calculation."""

    planet: str = Field(..., description="Planet name")
    julian_day: float = Field(..., description="Julian Day Number")
    position: PlanetPosition = Field(..., description="Position data")
    calculation_time: datetime = Field(
        ..., description="UTC time of calculation"
    )


class BatchCalculationRequest(BaseModel):
    """Request model for batch planetary position calculations."""

    calculations: List[CalculationRequest] = Field(
        ..., description="List of calculations to perform"
    )


class BatchCalculationResponse(BaseModel):
    """Response model for batch planetary position calculations."""

    results: List[CalculationResponse] = Field(
        ..., description="List of calculation results"
    )
    calculation_time: datetime = Field(
        ..., description="UTC time of batch calculation"
    )


class EphemerisHealthResponse(BaseModel):
    """Response model for ephemeris health check."""

    status: str = Field(..., description="Service status")
    timestamp: datetime = Field(..., description="Health check timestamp")
    ephemeris_initialized: bool = Field(
        ..., description="Whether ephemeris is properly initialized"
    )

File: tree-shaking-backup/backup-2025-08-27T03-27-57-791Z/packages/integrations/src/ephemeris.ts
/**
 * Ephemeris types and client utilities for shared use across applications.
 *
 * This module provides TypeScript types and utility functions for interacting
 * with the ephemeris server, ensuring consistency across astro and healwave apps.
 */

  /** Whether the planet is in retrograde motion */
  retrograde: boolean;
}

  /** Planet name (e.g., 'sun', 'moon', 'mercury') */
  planet: string;
}

  /** Julian Day Number */
  julian_day: number;
  /** Position data */
  position: PlanetPosition;
  /** UTC timestamp of calculation */
  calculation_time: string;
}

}

  /** UTC timestamp of batch calculation */
  calculation_time: string;
}

  /** Health check timestamp */
  timestamp: string;
  /** Whether ephemeris is properly initialized */
  ephemeris_initialized: boolean;
}

/**
 * Supported planetary bodies in the ephemeris system.
 */
export const SUPPORTED_PLANETS = [
  'sun',
  'moon',
  'mercury',
  'venus',
  'mars',
  'jupiter',
  'saturn',
  'uranus',
  'neptune',
  'pluto',
  'chiron',
  'ceres',
  'pallas',
  'juno',
  'vesta',
] as const;

/**
 * Configuration for ephemeris client.
 */
  /** API key for authentication (if required) */
  apiKey?: string;
  /** Request timeout in milliseconds */
  timeout?: number;
}

/**
 * Client for fetching ephemeris data from the backend API.
 *
 * This client abstracts the communication with the backend, which in turn
 * communicates with the dedicated ephemeris server.
 */

/**
 * Utility function to convert a Date to Julian Day Number.
 *
 * @param date - The date to convert
 * @returns Julian Day Number
 */
export function dateToJulianDay(date: Date): number {
  const a = Math.floor((14 - date.getMonth() - 1) / 12);
  const y = date.getFullYear() + 4800 - a;
  const m = date.getMonth() + 1 + 12 * a - 3;

  const jdn =
    date.getDate() +
    Math.floor((153 * m + 2) / 5) +
    365 * y +
    Math.floor(y / 4) -
    Math.floor(y / 100) +
    Math.floor(y / 400) -
    32045;

  // Add fractional day
  const hours = date.getHours();
  const minutes = date.getMinutes();
  const seconds = date.getSeconds();
  const milliseconds = date.getMilliseconds();

  const fractionalDay =
    (hours - 12) / 24 +
    minutes / 1440 +
    seconds / 86400 +
    milliseconds / 86400000;

  return jdn + fractionalDay;
}

/**
 * Utility function to convert Julian Day Number to Date.
 *
 * @param julianDay - Julian Day Number
 * @returns Date object
 */

/**
 * Utility function to format planet position for display.
 *
 * @param position - Planet position data
 * @param precision - Number of decimal places (default: 2)
 * @returns Formatted position string
 */

/**
 * Utility function to get astrological sign from degree position.
 *
 * @param degrees - Position in degrees (0-360)
 * @returns Astrological sign information
 */
export function getAstrologicalSign(degrees: number): {
  sign: string;
  signDegrees: number;
  signMinutes: number;
} {
  const signs = [
    'Aries',
    'Taurus',
    'Gemini',
    'Cancer',
    'Leo',
    'Virgo',
    'Libra',
    'Scorpio',
    'Sagittarius',
    'Capricorn',
    'Aquarius',
    'Pisces',
  ];

  const normalizedDegrees = degrees % 360;
  const signIndex = Math.floor(normalizedDegrees / 30);

  // Ensure signIndex is within valid bounds
  const clampedSignIndex = Math.max(0, Math.min(signIndex, signs.length - 1));
  const sign = signs[clampedSignIndex];

  if (!sign) {
    // Fallback to Aries if something goes wrong
    return {
      sign: 'Aries',
      signDegrees: 0,
      signMinutes: 0,
    };
  }

  const signDegrees = normalizedDegrees % 30;
  const signMinutes = (signDegrees % 1) * 60;

  return {
    sign,
    signDegrees: Math.floor(signDegrees),
    signMinutes: Math.floor(signMinutes),
  };
}

/**
 * Create a configured ephemeris client instance.
 *
 * @param config - Client configuration
 * @returns Configured ephemeris client
 */
export function createEphemerisClient(
  config: EphemerisConfig
): EphemerisClient {
  return new EphemerisClient(config);
}
